kong/clustering/services/negotiation.lua

```diff
@@ -0,0 +1,322 @@
+      service = version
+    }
+  end
+
+  return wrapped
+end
+
+local function unwrap_services(service)
+  local unwrapped = {}
+  for name, version in pairs(service) do
+    unwrapped[name] = version.service
+  end
+
+  return unwrapped
+end
+
+local _M = {}
+
+local function field_validate(tbl, field, typ)
+  local v = tbl
+  for i, ind in ipairs(field) do
+    if type(v) ~= "table" then
+      error("field '" .. table_concat(field, ".", 1, i - 1) .. "' cannot be indexed with " .. ind)
+    end
+    v = v[ind]
+  end
+
+  local compare_typ = typ
+  if typ == "array" or typ == "object" then
+    compare_typ = "table"
+  end
+
+  if type(v) ~= compare_typ then
+    local field_name = table_concat(field, '.')
+    error("field \"" .. field_name .. "\" must be of type " .. typ)
+  end
+end
+
+local function verify_request(body)
+  for field, typ in pairs{
+    [{
+      "node",
+    }] = "object",
+    [{
+      "node", "type",
+    }] = "string",
+    [{
+      "node", "version",
+    }] = "string",
+    [{
+      "services_requested",
+    }] = "array",
+  } do
+    field_validate(body, field, typ)
+  end
+end
+
+local function verify_node_compatibility(client_node)
+  if client_node.type ~= "KONG" then
+    error(("unknown node type %q"):format(client_node.type), CLUSTERING_SYNC_STATUS.UNKNOWN)
+  end
+
+  local ok, err, result = clustering_utils.check_kong_version_compatibility(kong.version, client_node.version)
+  if not ok then
+    error(err)
+  end
+  return result
+end
+
+local function negotiate_version(name, versions, known_versions)
+  local versions_set = {}
+  for _, v in ipairs(versions) do
+    versions_set[lower(v.version)] = true
+  end
+
+  for _, v in ipairs(known_versions) do
+    local version = lower(v.version)
+    if versions_set[version] then
+      return v
+    end
+  end
+
+  return { name = name, description = "No valid version" }
+end
+
+local function negotiate_service(name, versions)
+  name = lower(name)
+
+  if type(versions) ~= "table" then
+    error("invalid versions array for service " .. name)
+  end
+
+  local supported_service = supported_services[name]
+  if not supported_service then
+    return { description = "unknown service." }
+  end
+
+  return negotiate_version(name, versions, supported_service)
+end
+
+local function log_negotiation_result(name, version)
+  if version.version ~= nil then
+    log(NOTICE, "service accepted: \"", name, "\", version: ", version.version, ", description: ", version.description)
+
+  else
+    log(NOTICE, "service rejected: \"", name, "\", reason: ", version.description)
+  end
+end
+
+local function negotiate_services(services_requested)
+  local services = {}
+
+  for name, versions in pairs(services_requested) do
+    if type(name) ~= "string" or type(versions) ~= "table" then
+      error("malformed service requested " .. name)
+    end
+
+    local negotiated_version = negotiate_service(name, versions)
+    services[name] = negotiated_version
```

Comment:
this table is always traversed, never consulted.  it's better served by an array instead of a dictionary.